pycrostates/cluster/aahc.py

```diff
@@ -0,0 +1,386 @@
+
+    @copy_doc(_BaseCluster._check_fit)
+    def _check_fit(self):
+        super()._check_fit()
+        # sanity-check
+        assert self.GEV_ is not None
+
+    @copy_doc(_BaseCluster.fit)
+    @fill_doc
+    def fit(
+        self,
+        inst: Union[BaseRaw, BaseEpochs],
+        picks: Picks = "eeg",
+        tmin: Optional[Union[int, float]] = None,
+        tmax: Optional[Union[int, float]] = None,
+        reject_by_annotation: bool = True,
+        n_jobs: int = 1,
+        *,
+        verbose: Optional[str] = None,
+    ) -> None:
+        """
+        %(verbose)s
+        """
+        _set_verbose(verbose)  # TODO: decorator nesting is failing
+        data = super().fit(
+            inst, picks, tmin, tmax, reject_by_annotation, n_jobs
+        )
+
+        gev, maps, segmentation = AAHCluster._aahc(
+            data,
+            self._n_clusters,
+            self._ignore_polarity,
+            self._normalize_input,
+        )
+
+        if gev is not None:
+            logger.info("AAHC converged with GEV = %.2f%% ", gev * 100)
+
+        self._GEV_ = gev
+        self._cluster_centers_ = maps
+        self._labels_ = segmentation
+        self._fitted = True
+
+    @copy_doc(_BaseCluster.save)
+    def save(self, fname: Union[str, Path]):
+        super().save(fname)
+        # TODO: to be replaced by a general writer than infers the writer from
+        # the file extension.
+        from ..io.fiff import _write_cluster  # pylint: disable=C0415
+
+        _write_cluster(
+            fname,
+            self._cluster_centers_,
+            self._info,
+            "AAHCluster",
+            self._cluster_names,
+            self._fitted_data,
+            self._labels_,
+            ignore_polarity=self._ignore_polarity,
+            normalize_input=self._normalize_input,
+            GEV_=self._GEV_,
+        )
+
+    # --------------------------------------------------------------------
+    @staticmethod
+    def _aahc(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+    ) -> Tuple[float, NDArray[float], NDArray[int]]:
+        """Run the AAHC algorithm."""
+        gfp_sum_sq = np.sum(data**2)
+        maps, segmentation = AAHCluster._compute_maps(
+            data, n_clusters, ignore_polarity, normalize_input
+        )
+        map_corr = _corr_vectors(data, maps[segmentation].T)
+        gev = np.sum((data * map_corr) ** 2) / gfp_sum_sq
+        return gev, maps, segmentation
+
+    @staticmethod
+    def _compute_maps(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
```

Comment:
```suggestion
        cluster = data.copy()

        if normalize_input:
            cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
```

Shouldn't it be this? Or am I missing something?

Reply:
The version that I submitted is consistent with the reference implementation [1] (eeg_computeAAHC.m lines 15 to 19). That is, the clusters *and* data are normalized.

[1] https://www.thomaskoenig.ch/Download/EEGLAB_Microstates/Microstates1.2.zip

Reply:
I completely misread those lines, I thought this copy wasn't achieving anything, but it does replace the argument `data` with itself normalized.